pkg/blobstore/spanner_blob_access.go

```diff
@@ -0,0 +1,975 @@
+			Help:      "Number of failed deletes of GCS blobs that couldn't be read",
+		})
+	gcsPutFailedContextCanceledCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: "buildbarn",
+			Subsystem: "blobstore",
+			Name:      "gcs_put_failed_context_canceled_total",
+			Help:      "Number of failed puts of GCS blobs because the context was canceled",
+		})
+	gcsPutFailedDeadlineExceededCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: "buildbarn",
+			Subsystem: "blobstore",
+			Name:      "gcs_put_failed_deadline_exceeded_total",
+			Help:      "Number of failed puts of GCS blobs because the deadline was exceeded",
+		})
+	gcsPutFailedOtherCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Namespace: "buildbarn",
+			Subsystem: "blobstore",
+			Name:      "gcs_put_failed_other_total",
+			Help:      "Number of failed puts of GCS blobs because of other reasons",
+		})
+	spannerMalformedBlobDeletedCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "buildbarn",
+			Subsystem: "blobstore",
+			Name:      "spanner_malformed_blob_deleted_total",
+			Help:      "Number of malformed blobs that were deleted",
+		},
+		[]string{"backend_type"})
+	spannerMalformedBlobDeleteFailedCount = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "buildbarn",
+			Subsystem: "blobstore",
+			Name:      "spanner_malformed_blob_delete_failed_total",
+			Help:      "Number of malformed blobs that could not be deleted",
+		},
+		[]string{"backend_type"})
+	backendOperationsDurationSeconds = prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Namespace: "buildbarn",
+			Subsystem: "blobstore",
+			Name:      "backend_operations_duration_seconds",
+			Help:      "Amount of time spent per backend operation in seconds.",
+			Buckets:   util.DecimalExponentialBuckets(-3, 6, 2),
+		},
+		[]string{"storage_type", "backend_type", "operation"})
+
+	BEMap = map[pb.StorageType_Value]string{
+		pb.StorageType_CASTORE:      ST_CAS,
+		pb.StorageType_ACTION_CACHE: ST_AC,
+	}
+)
+
+type spannerBlobAccess struct {
+	capabilities.Provider
+
+	spannerClient *spanner.Client
+	gcsBucket     *storage.BucketHandle
+
+	readBufferFactory ReadBufferFactory
+	storageType       pb.StorageType_Value
+	daysToLive        uint64        // to avoid converting back and forth
+	expirationAge     time.Duration // same as above, but easier for time calculations
+	refUpdateThresh   time.Duration // when we start updating ReferenceTime
+	refChan           chan string
+}
+
+type spannerRecord struct {
+	Key           string
+	InlineData    []byte
+	ReferenceTime time.Time
+}
+
+// databaseName is of the form "projects/<project ID>/instances/<instance name>/databases/<database name>".
+// tableName should include the name of the table and the version of the table's schema.
+func createSpannerTable(ctx context.Context, databaseName string, daysToLive uint64) error {
```

Comment:
Do we think it's a good idea to have the table creation living here? Doesn't this require that your processes not only have query/insert/delete rights on the table, but also administrative rights to create new tables? I'd rather see that we provide some SQL file for people to run manually when setting up their cluster. Just like how we also don't auto-create S3 buckets.

Reply:
Each of our services run in separate projects with separate permissions.  The table is BuildBarn's, so it naturally creates it if it is missing.  Our Tech Risk team didn't see anything wrong with this approach.  Humans can, of course, create the table manually, but that is error prone.  That's why we automated it.